### PR TITLE
Improve metrics queries

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
This PR updates the metrics query to use a recursive query for recomputing the periods and then joining the metrics with the periods.
That has the advantage of getting rid of the integer rounding and casting we do right now to determine the period for each metric. Instead periods are essentially precomputed and metrics are then matched to periods.

### CODE REVIEW

File: api/metrics.go [hunk 1]
```diff
@@ -1,12 +1,19 @@
 package api
 
 import (
+	"fmt"
```
Comment (ChrisSchinnerl) [inline api/metrics.go:4]: Blocker to test this with the UI

File: stores/metrics.go [hunk 2]
```diff
@@ -517,36 +517,49 @@ func normaliseTimestamp(start time.Time, interval time.Duration, t unixTimeMS) u
...
-	if !isSQLite(db) {
-		return gorm.Expr("CAST(FLOOR((timestamp - ?) / ?) * ? AS SIGNED)", unixTimeMS(start), interval.Milliseconds(), interval.Milliseconds())
-	} else {
-		return gorm.Expr("(timestamp - ?) / ? * ?", unixTimeMS(start), interval.Milliseconds(), interval.Milliseconds())
-	}
-}
-
 func (s *SQLStore) findAggregatedContractPeriods(start time.Time, n uint64, interval time.Duration) ([]dbContractMetric, error) {
+	if n > api.MaxMetricIntervals {
+		return nil, api.ErrMaxIntervalsExceeded
```
Comment (peterjan) [inline stores/metrics.go:522]: handlers should catch and return a 404

File: stores/metrics.go [hunk 3]
```diff
@@ -567,29 +580,47 @@ func (s *SQLStore) findAggregatedContractPeriods(start time.Time, n uint64, inte
 // split into intervals and the row with the lowest timestamp for each interval
 // is returned. The result is then joined with the original table to retrieve
 // only the metrics we want.
-func (s *SQLStore) findPeriods(tx *gorm.DB, dst interface{}, start time.Time, n uint64, interval time.Duration) error {
+func (s *SQLStore) findPeriods(table string, dst interface{}, start time.Time, n uint64, interval time.Duration, whereExpr clause.Expr) error {
+	if n > api.MaxMetricIntervals {
+		return api.ErrMaxIntervalsExceeded
```
Comment (peterjan) [inline stores/metrics.go:585]: handlers should catch and return a 404

File: api/metrics.go [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
 
 	"go.sia.tech/core/types"
 )
 
+var (
+	ErrMaxIntervalsExceeded = fmt.Errorf("max number of intervals exceeds maximum of %v", MaxMetricIntervals)
+)
+
 const (
+	MaxMetricIntervals = 1000
```
Comment (peterjan) [inline api/metrics.go:15]: `uber nit` `MetricMaxIntervals`

File: stores/metrics.go [hunk 5]
```diff
@@ -517,36 +517,49 @@ func normaliseTimestamp(start time.Time, interval time.Duration, t unixTimeMS) u
...
+		FROM
+			periods p
+		INNER JOIN
+			contracts c ON c.timestamp >= p.period_start AND c.timestamp < p.period_start + ?
+		GROUP BY
+			p.period_start, c.fcid
+		ORDER BY
+			p.period_start ASC
+		) i ON contracts.id = i.id
+	`, unixTimeMS(start),
```
Comment (peterjan) [inline stores/metrics.go:551]: Is the period start rounded down to the nearest interval? Seems you'd get different periods if you compare to what we used to do in `roundPeriodExpr`

File: api/metrics.go (hunk 1 above)
Comment (ChrisSchinnerl) [reply api/metrics.go:4]: Confirmed that it's working.

File: stores/metrics.go (hunk 5 above)
Comment (ChrisSchinnerl) [reply stores/metrics.go:551]: The periods don't in the query don't really matter as long as they are distinct and contracts are associated correctly with them. `normaliseTimestamp` later updates the timestamp of each returned metric to be correct.

File: stores/metrics.go (hunk 2 above)
Comment (ChrisSchinnerl) [reply stores/metrics.go:522]: Good point. But I went for `BadRequest` instead of `NotFound`.

Comment (peterjan) [reply stores/metrics.go:522]: lol

**Review by peterjan — APPROVED** [review]

**Pull request merged**